// tailwindcss/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.simpleConfigStubFile = exports.defaultPostCssConfigFile = exports.defaultConfigStubFile = exports.supportedPostCssConfigFile = exports.cjsConfigFile = exports.supportedConfigFiles = exports.defaultPostCssConfigStubFile = exports.cli = exports.cjsPostCssConfigFile = exports.defaultConfigFile = void 0;
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const cli = 'tailwind';
exports.cli = cli;
const defaultConfigFile = './tailwind.config.js';
exports.defaultConfigFile = defaultConfigFile;
const defaultPostCssConfigFile = './postcss.config.js';
exports.defaultPostCssConfigFile = defaultPostCssConfigFile;
const cjsConfigFile = './tailwind.config.cjs';
exports.cjsConfigFile = cjsConfigFile;
const cjsPostCssConfigFile = './postcss.config.cjs';
exports.cjsPostCssConfigFile = cjsPostCssConfigFile;
const supportedConfigFiles = [
    cjsConfigFile,
    defaultConfigFile
];
exports.supportedConfigFiles = supportedConfigFiles;
const supportedPostCssConfigFile = [
    cjsPostCssConfigFile,
    defaultPostCssConfigFile
];
exports.supportedPostCssConfigFile = supportedPostCssConfigFile;
const defaultConfigStubFile = _path.default.resolve(__dirname, '../stubs/defaultConfig.stub.js');
exports.defaultConfigStubFile = defaultConfigStubFile;
const simpleConfigStubFile = _path.default.resolve(__dirname, '../stubs/simpleConfig.stub.js');
exports.simpleConfigStubFile = simpleConfigStubFile;
const defaultPostCssConfigStubFile = _path.default.resolve(__dirname, '../stubs/defaultPostCssConfig.stub.js');
exports.defaultPostCssConfigStubFile = defaultPostCssConfigStubFile;
